[processor/tailsamplingprocessor] Add `dropped` attribute to metrics tracking sampling decisions

yvrhdn opened the pull request:
#### Description

The tail sampling processor has a few metrics to track sampling decision per policy:

```
otelcol_processor_tail_sampling_count_traces_sampled
otelcol_processor_tail_sampling_count_spans_sampled
otelcol_processor_tail_sampling_global_count_traces_sampled
```

These metrics have an attribute `sampled` which is true or false depending on the sampling decision.

With the introduction of drop policies (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39668) this has become more nuanced. `sampled="false"` can mean two things:
* none of the policies matched and the trace was not sampled
* a drop policy matched and the trace was discarded without evaluating any other policy

It's currently not possible to distinguish between "not sampled by any policy" and "actively dropped" traces. This PR adds a new attribute `dropped` which can be used to track traces actively dropped.

#### Testing
Tests have been updated to check the new attribute.

#### Documentation
Added an example query to show frequency of dropped traces.

yvrhdn committed:
[processor/tailsamplingprocessor] Add `dropped` attribute to metrics tracking sampling decisions

yvrhdn committed:
add example query

yvrhdn committed:
add tsp-telemetry-add-dropped-attribute.yaml

yvrhdn committed:
Merge branch 'main' into y/telemetry-add-dropped-attribute

atoulme commented:
@portertech please take a look

portertech reviewed (changes requested):
What about a test with dropped `true`?

yvrhdn committed:
Add test `TestMetricsCountSampled` with sample and drop test case

I've renamed `TestProcessorTailSamplingCountSpansSampled` to `TestMetricsCountSampled` and extended it to test all metrics with `count_{unit}_sampled`:
- `otelcol_processor_tail_sampling_global_count_traces_sampled`
- `otelcol_processor_tail_sampling_count_traces_sampled`
- `otelcol_processor_tail_sampling_count_spans_sampled` (was already in this test)

To fix `global_count_traces_sampled` I now let `makeDecision` also return `Dropped` instead of only `Sampled` or `NotSampled`.

yvrhdn committed:
Merge branch 'main' into y/telemetry-add-dropped-attribute

TylerHelmuth commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -77,14 +77,15 @@ type spanAndScope struct {
-	attrSampledTrue     = metric.WithAttributes(attribute.String("sampled", "true"))
-	attrSampledFalse    = metric.WithAttributes(attribute.String("sampled", "false"))
-	decisionToAttribute = map[sampling.Decision]metric.MeasurementOption{
+	attrSampledTrue      = metric.WithAttributes(attribute.String("sampled", "true"), attribute.String("dropped", "false"))
+	attrSampledFalse     = metric.WithAttributes(attribute.String("sampled", "false"), attribute.String("dropped", "false"))
+	attrDropped          = metric.WithAttributes(attribute.String("sampled", "false"), attribute.String("dropped", "true"))
```
Attributes on emitted metrics should be included in the metadata.yaml like https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1f92d42bea48e3fcbb351f95340563990001bd2a/receiver/kubeletstatsreceiver/metadata.yaml#L73-L81. Can you update it with the `sampled` and `dropped` attributes?

yvrhdn replied:
Ah, I wasn't aware this exists, thanks!

yvrhdn committed:
document attributes in metadata.yaml

yvrhdn committed:
Merge branch 'main' into y/telemetry-add-dropped-attribute

yvrhdn committed:
document attributes
